extension/src/codeLensProvider.ts

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+import { CancellationToken, CodeLens, CodeLensProvider, Event, EventEmitter, TextDocument } from 'vscode';
+import { JavaTestRunnerCommands } from './constants/commands';
+import { ITestItem } from './protocols';
+import { searchTestCodeLens } from './utils/commandUtils';
+
+class TestCodeLensProvider implements CodeLensProvider {
+    private _onDidChangeCodeLenses: EventEmitter<void> = new EventEmitter<void>();
+
+    get onDidChangeCodeLenses(): Event<void> {
+        return this._onDidChangeCodeLenses.event;
+    }
+
+    public refresh(): void {
+        this._onDidChangeCodeLenses.fire();
+    }
+
+    public async provideCodeLenses(document: TextDocument, token: CancellationToken): Promise<CodeLens[]> {
+        try {
+            const testItems: ITestItem[] = await searchTestCodeLens(document.uri.toString());
+            const codeLenses: CodeLens[] = [];
+            for (const test of testItems) {
+                if (token.isCancellationRequested) {
```

Comment:
The heavy job is searching code lens, can you put the cancellation logic inside searchTestCodeLens?